reasoner/resolution/resolver/DisjunctionResolver.java

```diff
@@ -0,0 +1,220 @@
+                nextAnswer(fromUpstream, responseProducer, iteration);
+                return;
+            }
+            answerToUpstream(answer, fromUpstream, iteration);
+        } else {
+            nextAnswer(fromUpstream, responseProducer, iteration);
+        }
+    }
+
+    @Override
+    protected void receiveFail(Response.Fail fromDownstream, int iteration) {
+        LOG.trace("{}: received Exhausted, with iter {}: {}", name(), iteration, fromDownstream);
+        Request toDownstream = fromDownstream.sourceRequest();
+        Request fromUpstream = fromUpstream(toDownstream);
+        ResponseProducer responseProducer = responseProducers.get(fromUpstream);
+
+        if (iteration < responseProducer.iteration()) {
+            // short circuit old iteration failed messages back out of the actor model
+            failToUpstream(fromUpstream, iteration);
+            return;
+        }
+        responseProducer.removeDownstreamProducer(fromDownstream.sourceRequest());
+        nextAnswer(fromUpstream, responseProducer, iteration);
+    }
+
+    @Override
+    protected void initialiseDownstreamResolvers() {
+        LOG.debug("{}: initialising downstream resolvers", name());
+        for (grakn.core.pattern.Conjunction conjunction : disjunction.conjunctions()) {
+            downstreamResolvers.add(registry.nested(conjunction));
+        }
+        isInitialised = true;
+    }
+
+    private ResponseProducer mayUpdateAndGetResponseProducer(Request fromUpstream, int iteration) {
+        if (!responseProducers.containsKey(fromUpstream)) {
+            responseProducers.put(fromUpstream, responseProducerCreate(fromUpstream, iteration));
+        } else {
+            ResponseProducer responseProducer = responseProducers.get(fromUpstream);
+            assert responseProducer.iteration() == iteration ||
+                    responseProducer.iteration() + 1 == iteration;
+
+            if (responseProducer.iteration() + 1 == iteration) {
+                // when the same request for the next iteration the first time, re-initialise required state
+                ResponseProducer responseProducerNextIter = responseProducerReiterate(fromUpstream, responseProducer, iteration);
+                responseProducers.put(fromUpstream, responseProducerNextIter);
+            }
+        }
+        return responseProducers.get(fromUpstream);
+    }
+
+    @Override
+    protected ResponseProducer responseProducerReiterate(Request fromUpstream, ResponseProducer responseProducerPrevious,
+                                                         int newIteration) {
+        assert newIteration > responseProducerPrevious.iteration();
+        LOG.debug("{}: Updating ResponseProducer for iteration '{}'", name(), newIteration);
+
+        assert newIteration > responseProducerPrevious.iteration();
+        ResponseProducer responseProducerNewIter = responseProducerPrevious.newIterationRetainDedup(Iterators.empty(), newIteration);
```

Comment:
we don't actually want to retain dedup sets in non-root response producers!! this should be non-dedup and the `Root.Disjunction` should be using dedup